### PR TITLE
Provide inboard/outboard differentiation.

### PR DESCRIPTION
**Pull request by connoramoreno**
Changes "tokamak.py" to allow differentiation between inboard and outboard in the tokamak radial build. Also provides a more elaborate and realistic example radial build.

### CODE REVIEW

File: tokamak.py [hunk 1]
```diff
@@ -21,29 +23,80 @@ def read_radial_build(file):
...
+
+		inboard_id = len(inboard_thicknesses) + 4
+		for id, layer_name in zip(range(inboard_id-2,2,-1), reversed(inboard_layers)):
+			cubit.cmd("subtract volume "+str(id - 1)+" from volume "+str(id)+" keep_tool")
+			cubit.cmd("intersect vol 1 "+str(inboard_id-1)+" keep")
+			cubit.cmd("group 'mat:"+ layer_name +"' add vol " + str(inboard_id))
+			cubit.cmd("delete vol "+str(inboard_id-1))
+			inboard_id += 2
+
+		outboard_radii = np.cumsum(outboard_thicknesses) + minor_radius
```
Comment (gonuke) [inline tokamak.py]: There seems to be an opportunity for modularity here, with the O/B being nearly the same as the I/B

File: tokamak.py [hunk 2]
```diff
@@ -21,29 +23,80 @@ def read_radial_build(file):
...
-	cubit.cmd("group 'mat:Vacuum' add vol 1")
-	vol_id = len(thicknesses) + 2
-	for id, layer_name in zip(range(vol_id-1,1,-1), reversed(layers)):
-		cubit.cmd("subtract volume "+str(id - 1)+" from volume "+str(id)+" keep_tool")
-		cubit.cmd("group 'mat:"+ layer_name +"' add vol " + str(vol_id))
-		vol_id += 1
-
-	cubit.cmd("imprint volume all")
-	cubit.cmd("merge volume all")
+		radii = np.cumsum(thicknesses) + minor_radius
```
Comment (gonuke) [inline tokamak.py]: Possible reuse/modularity here, as well

**Review by gonuke — CHANGES_REQUESTED** [review]
There are now three blocks of nearly identical code: one for I/B, one for O/B, and one for symmetric.  Consider making a function that does this with some options, depending on which case it is.

Also, how do we expect the I/B & O/B to meet at the top/bottom?

File: exampleradialbuild.yaml [hunk 3]
```diff
@@ -1,6 +1,20 @@
...
+  tf_magnets: 63
+
+Outboard:
+  first_wall: 4
+  breeder: 100
+  manifold: 6
+  structural_ring: 20
+  vacuum_vessel: 10
+  lt_shield: 17
+  tf_magnets: 63
```
Comment (ypark234) [inline exampleradialbuild.yaml]: > it is a good style to always put the newline as a last character

https://stackoverflow.com/questions/5813311/no-newline-at-end-of-file

**Review by ubaker — COMMENTED** [review]
Test comment

File: tokamak.py [hunk 4]
```diff
@@ -21,29 +23,80 @@ def read_radial_build(file):
...
+		inboard_id = len(inboard_thicknesses) + 4
+		for id, layer_name in zip(range(inboard_id-2,2,-1), reversed(inboard_layers)):
+			cubit.cmd("subtract volume "+str(id - 1)+" from volume "+str(id)+" keep_tool")
+			cubit.cmd("intersect vol 1 "+str(inboard_id-1)+" keep")
+			cubit.cmd("group 'mat:"+ layer_name +"' add vol " + str(inboard_id))
+			cubit.cmd("delete vol "+str(inboard_id-1))
+			inboard_id += 2
+
+		outboard_radii = np.cumsum(outboard_thicknesses) + minor_radius
+		for radius in outboard_radii:
```
Comment (opotowsky) [inline tokamak.py:56]: test comment

File: exampleradialbuild.yaml (hunk 3 above)
Comment (ubaker) [inline exampleradialbuild.yaml]: Another test comment!

File: tokamak.py [hunk 5]
```diff
@@ -13,6 +14,7 @@ def parse_args():
 	args = parser.parse_args()
 	return args
 
+
 def read_radial_build(file):
```
Comment (ypark234) [inline tokamak.py]: Docstring might be useful for readers of the code.
https://realpython.com/documenting-python-code/#documenting-your-python-code-base-using-docstrings

File: exampleradialbuild.yaml (hunk 3 above)
Comment (bam241) [inline exampleradialbuild.yaml]: we probably want documentation about all this,
At least to get a list of which ones are required, and which ones are optional, what they mean and if there is default values

File: tokamak.py [hunk 6]
```diff
@@ -3,6 +3,7 @@
 import yaml
 import numpy as np
 
+
 def parse_args():
```
Comment (bam241) [inline tokamak.py]: PEP8: empty line not required

File: tokamak.py [hunk 7]
```diff
@@ -21,29 +23,80 @@ def read_radial_build(file):
 
 def build_torus(major_radius,minor_radius,radial_build):
```
Comment (bam241) [inline tokamak.py]: pep8: no comments

File: tokamak.py [hunk 8]
```diff
@@ -21,29 +23,80 @@ def read_radial_build(file):
...
```
Comment (bam241) [inline tokamak.py]: pep8: space after `,`

File: tokamak.py [hunk 9]
```diff
@@ -21,29 +23,80 @@ def read_radial_build(file):
...
+			cubit.cmd("delete vol "+str(inboard_id-1))
+			inboard_id += 2
+
+		outboard_radii = np.cumsum(outboard_thicknesses) + minor_radius
+		for radius in outboard_radii:
+			cubit.torus(major_radius,radius)
+
+		outboard_id = inboard_id + len(outboard_thicknesses) - 1
+		for id, layer_name in zip(range(outboard_id-1,inboard_id-1,-1), reversed(outboard_layers)):
+					cubit.cmd("subtract volume "+str(id-1)+" from volume "+str(id)+" keep_tool")
```
Comment (opotowsky) [inline tokamak.py]: Looks like indentation is a little off here?

File: tokamak.py [hunk 10]
```diff
@@ -21,29 +23,80 @@ def read_radial_build(file):
...
+		cubit.cylinder(4*major_radius,major_radius,major_radius,major_radius)
+
+		cubit.torus(major_radius,minor_radius)
+		cubit.cmd("group 'mat:Vacuum' add vol 2")
+
+		inboard_radii = np.cumsum(inboard_thicknesses) + minor_radius
+		for radius in inboard_radii:
+			cubit.torus(major_radius,radius)
+
+		inboard_id = len(inboard_thicknesses) + 4
```
Comment (bam241) [inline tokamak.py]: in general it is good to avoid magic number. (here the `4`)

File: tokamak.py [hunk 11]
```diff
@@ -21,29 +23,80 @@ def read_radial_build(file):
...
+		for id, layer_name in zip(range(inboard_id-2,2,-1), reversed(inboard_layers)):
+			cubit.cmd("subtract volume "+str(id - 1)+" from volume "+str(id)+" keep_tool")
+			cubit.cmd("intersect vol 1 "+str(inboard_id-1)+" keep")
+			cubit.cmd("group 'mat:"+ layer_name +"' add vol " + str(inboard_id))
+			cubit.cmd("delete vol "+str(inboard_id-1))
```
Comment (bam241) [inline tokamak.py]: pep8 recommends spaces before & after math signs (`=`, `+`,...)

File: tokamak.py [hunk 12]
```diff
@@ -3,6 +3,7 @@
 import yaml
 import numpy as np
 
+
 def parse_args():
```
Comment (ypark234) [inline tokamak.py]: I'd suggest renaming this to something slightly different, as it has the same name as `parse_args` method of `argparse`, as in line 14. (It probably won't make any problem here as namespace is different, but just for good? practice.)

File: tokamak.py [hunk 13]
```diff
@@ -21,29 +23,80 @@ def read_radial_build(file):
...
+		outboard_radii = np.cumsum(outboard_thicknesses) + minor_radius
+		for radius in outboard_radii:
+			cubit.torus(major_radius,radius)
+
+		outboard_id = inboard_id + len(outboard_thicknesses) - 1
+		for id, layer_name in zip(range(outboard_id-1,inboard_id-1,-1), reversed(outboard_layers)):
+					cubit.cmd("subtract volume "+str(id-1)+" from volume "+str(id)+" keep_tool")
+					cubit.cmd("subtract vol 1 from vol "+str(outboard_id)+" keep_tool")
+					cubit.cmd("group 'mat:"+ layer_name +"' add vol " + str(outboard_id))
+					outboard_id += 1
```
Comment (bam241) [inline tokamak.py]: pep8: missing spaces around math signs

**Review by bam241 — CHANGES_REQUESTED** [review]
Super job  some few comments mainly on styles.

avoid using tab, prefers space for indentation, 4 spaces per indent is usually recommended for python

also in general it is recommend to add docstring to explain what each methods is about, what are the arguments, and what it returns.

I think you are a little over using empty line, it is good to space the code, but keeping things together also helps readability.

File: tokamak.py [hunk 14]
```diff
@@ -13,6 +14,7 @@ def parse_args():
 	args = parser.parse_args()
 	return args
 
+
 def read_radial_build(file):
 	
 	radial_build = yaml.load(file, Loader=yaml.FullLoader)
```
Comment (ypark234) [inline tokamak.py]: Variable name `file` is used by python as a keyword. I'd suggest changing it to `file_`
https://dbader.org/blog/meaning-of-underscores-in-python

**Review by ypark234 — CHANGES_REQUESTED** [review]
Good job, @connoramoreno ! A few comments.

**Comment by psbritt** [comment]
Unit test?

**Comment by stompsjo** [comment]
> Unit test?

Sounds like a good job for GitHub Actions 😉

**Comment by gonuke** [comment]
Note that by copying the file to a new name & deleting the old file, you have broken the commit history of the file.

Try `git mv old-filename new-filename` instead (in the future)

File: exampleradialbuild.yaml [hunk 15]
```diff
@@ -1,6 +1,27 @@
...
+# list represents a tokamak that is radially symmetric.
+
+# The example given below differentiates between the inboard and outboard.
+
+# Format
+# ------
+# <region>:
+#   <material_layer>: <thickness [cm]>
 ---
-fw: 5
```
Comment (gonuke) [inline exampleradialbuild.yaml:4]: Don't you want to keep some example of this structure as well?  It is still supported in your code, so you probably want something for testing, as well.

**Review by gonuke — CHANGES_REQUESTED** [review]
One comment on sample/test files.

File: exampleradialbuild.yaml (hunk 15 above)
Comment (connoramoreno) [reply exampleradialbuild.yaml:4]: good point! what do you think about inserting an example symmetric build just as comments above "Inboard"? could also just include a separate yaml file. does your expertise say one option is better than the other?

Comment (gonuke) [reply exampleradialbuild.yaml:4]: Recommend a separate yml file

**Comment by gonuke** [comment]
I think we're just waiting for a separate sample input with no IB/OB separation, right?

**Comment by connoramoreno** [comment]
yes, sorry was having git issues. also changed the script to allow repeated material names

**Comment by gonuke** [comment]
Note: best practice would be to make that a separate feature in a new pull request.  That way, the reviewed code can be merged and not prolong the PR.

File: MTOT.py [hunk 16]
```diff
@@ -0,0 +1,260 @@
...
+    cubit.cmd("group 'mat:Vacuum' add vol " + str(plasma_id))
+
+    if 'Inboard' in radial_build:
+        
+        inboard_layers = []
+        inboard_thicknesses = []
+        for index in range(len(radial_build['Inboard'])):
+            pair = radial_build['Inboard'][index]
+            inboard_layers += list(pair.keys())
+            inboard_thicknesses += list(pair.values())
```
Comment (gonuke) [inline MTOT.py:190]:
```suggestion
        for layer_info in radial_build['Inboard']:
            [(layer, thickness)] = layer_info.items()
            inboard_layers.append(layer)
            inboard_thicknesses.append(thickness)
```

File: MTOT.py [hunk 17]
```diff
@@ -0,0 +1,260 @@
...
+        vol_id = assemble_layers(vol_id, plasma_id, inboard_layers, cyl_id, "inboard")
+
+        first_outboard_id = vol_id + 1
+        
+        outboard_layers = []
+        outboard_thicknesses = []
+        for index in range(len(radial_build['Outboard'])):
+            pair = radial_build['Outboard'][index]
+            outboard_layers += list(pair.keys())
+            outboard_thicknesses += list(pair.values())
```
Comment (gonuke) [inline MTOT.py:203]: see above for inboard

File: MTOT.py [hunk 18]
```diff
@@ -0,0 +1,260 @@
...
+        
+        cubit.cmd("delete vol " + str(cyl_id))
+
+    else:
+        layers = []
+        thicknesses = []
+        for index in range(len(radial_build)):
+            pair = radial_build[index]
+            layers += list(pair.keys())
+            thicknesses += list(pair.values())
```
Comment (gonuke) [inline MTOT.py:222]: see above for inboard

Probably should be a function to unpack layers & thicknesses and reuse it in all three cases

**Review by gonuke — CHANGES_REQUESTED** [review]
Some feedback in the repeated layers addition.... and also need another example that demonstrates the repeated layers

**Pull request closed without merging**